graph/boruvka.cpp

```diff
@@ -0,0 +1,191 @@
+/**
+ * @file boruvkas_algorithm.cpp
+ * @brief Implementation of Boruvka's Algorithm to find out the minimum spanning
+ * tree of a graph.
+ * @author FranciscoThiesen
+ * @details For more details, see
+ * https://en.wikipedia.org/wiki/Bor%C5%AFvka%27s_algorithm
+ */
+
+#include <cassert>
+#include <iostream>
+#include <numeric>
+#include <vector>
+
+/**
+ * @brief Union-Find data structure (with union by rank and path compression).
+ */
+struct union_find {
+    std::vector<int> parent, component_size;
+
+    /**
+     * @brief Initializes the union_find data structure.
+     * @param n Number of elements.
+     */
+    union_find(int n) {
+        parent.resize(n);
+        component_size.assign(n, 1);
+        std::iota(parent.begin(), parent.end(), 0);
+    }
+
+    /**
+     * @brief Finds the representative of the set containing x.
+     * @param x An element of the set.
+     * @return The representative of the set containing x.
+     */
+    int find(int x) {
+        return (parent[x] == x) ? x : parent[x] = find(parent[x]);
+    }
+
+    /**
+     * @brief Unites two sets containing x and y.
+     * @param x An element of the first set.
+     * @param y An element of the second set.
+     */
+    void unite(int x, int y) {
+        x = find(x), y = find(y);
+        if (x != y) {
+            if (component_size[x] < component_size[y]) {
+                std::swap(x, y);
+            }
+            parent[y] = x;
+            component_size[x] += component_size[y];
+        }
+    }
+};
+
+/**
+ * @brief Boruvka's algorithm to compute the minimum spanning tree cost.
+ * @param num_nodes Number of nodes in the graph.
+ * @param num_edges Number of edges in the graph.
+ * @param edges Edges of the graph.
+ * @return Minimum spanning tree cost.
+ */
+long long boruvka(
+    int num_nodes, int num_edges,
```

Comment:
wouldnt it make more sense to make these unsigned since num_nodes/num_edges would never be negative?

Reply:
Will change it. Thanks for pointing this out.

Reply:
You are absolutely right! Just changed it.